mlir-sycl/lib/Conversion/SYCLToGPU/SYCLToGPU.cpp

```diff
@@ -0,0 +1,225 @@
+//===- SYCLToGPU.cpp - SYCL to GPU Patterns -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements patterns to convert SYCL dialect to GPU dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/SYCLToGPU/SYCLToGPU.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "mlir/Support/LLVM.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+#define DEBUG_TYPE "sycl-to-gpu"
+
+using namespace mlir;
+using namespace mlir::sycl;
+
+namespace {
+/// Returns the result of creating an operation to get a reference to an element
+/// of a sycl::id or sycl::range.
+Value createGetOp(OpBuilder &builder, Location loc, Type dimMtTy, Value res,
+                  Value index, ArrayAttr argumentTypes,
+                  FlatSymbolRefAttr functionName) {
+  return TypeSwitch<Type, Value>(res.getType())
+      .Case<IDType, RangeType>([&](auto arg) {
+        // `this` type
+        using ArgTy = decltype(arg);
+        // Operation type depending on ArgTy
+        using OpTy = std::conditional_t<std::is_same_v<ArgTy, IDType>,
+                                        SYCLIDGetOp, SYCLRangeGetOp>;
+        return builder.create<OpTy>(
+            loc, dimMtTy, res, index, argumentTypes, functionName, functionName,
+            builder.getAttr<FlatSymbolRefAttr>(ArgTy::getMnemonic()));
+      });
```

Comment:
We cannot currently lower this as is. I'll work on changing the lowering for `SYCLMethodOp`s in the following days.

Reply:
We shouldn't worry anyway, as we are not generating the operations we're lowering here anyway

Reply:
It's because of the dummy mangled function name